File: examples/kotlin/apps/portal/root.kt
package app.protocol.examples.kotlin.apps.portal

import app.protocol.examples.kotlin.cases.tasks.task_create.Task
import app.protocol.examples.kotlin.cases.tasks.task_create.TaskCreateUi
import app.protocol.examples.kotlin.cases.tasks.task_list.TaskListUi
import app.protocol.examples.kotlin.cases.tasks.task_move.TaskMoveUi
import app.protocol.examples.kotlin.core.ui.UiContext
import app.protocol.examples.kotlin.packages.design_system.DesignSystem
import org.w3c.dom.HTMLElement

object PortalRoot {
    fun render(
        registry: PortalRegistry,
        createUiContext: (Map<String, Any?>) -> UiContext,
        rootElement: HTMLElement,
    ) {
        var refreshToken = 0

        fun rerender() {
            val taskCreateContext =
                createUiContext(
                    mapOf(
                        "onTaskCreated" to { _: Task ->
                            refreshToken += 1
                            rerender()
                        },
                    ),
                )

            val taskListContext =
                createUiContext(
                    mapOf(
                        "refreshToken" to refreshToken,
                        "renderCardActions" to { task: app.protocol.examples.kotlin.cases.tasks.task_list.Task ->
                            val moveContext =
                                createUiContext(
                                    mapOf(
                                        "task" to app.protocol.examples.kotlin.cases.tasks.task_move.Task(
                                            id = task.id,
                                            title = task.title,
                                            description = task.description,
                                            status = task.status,
                                            createdAt = task.createdAt,
                                            updatedAt = task.updatedAt,
                                        ),
                                        "onTaskMoved" to { _: app.protocol.examples.kotlin.cases.tasks.task_move.Task ->
                                            refreshToken += 1
                                            rerender()
                                        },
                                    ),
                                )
                            TaskMoveUi(moveContext).view() as HTMLElement
                        },
                    ),
                )

            val taskCreateView =
                (registry._cases["tasks"]?.get("task_create")?.ui?.invoke(taskCreateContext)
                    as? TaskCreateUi
                    ?: TaskCreateUi(taskCreateContext))
                    .view() as HTMLElement
            val taskListView =
                (registry._cases["tasks"]?.get("task_list")?.ui?.invoke(taskListContext)
                    as? TaskListUi
                    ?: TaskListUi(taskListContext))
                    .view() as HTMLElement

            rootElement.innerHTML = ""
            rootElement.appendChild(
                DesignSystem.appShell(
                    title = "Task Board",
                    subtitle = "Kotlin APP example with create, list, and move wired through Cases.",
                    children = listOf(
                        DesignSystem.boardHeader(
                            title = "Board",
                            subtitle = "Tasks load from the backend and each card can move across columns.",
                        ),
                        taskCreateView,
                        taskListView,
                    ),
                ),
            )
        }

        rerender()
    }
}

File: examples/node/apps/portal/root.ts
import type { UiContext } from "../../core/ui.case";
import type { PortalRegistry } from "./registry";

interface PortalTask {
  id: string;
  title: string;
  description?: string;
  status: "todo" | "doing" | "done";
  createdAt: string;
  updatedAt: string;
}

interface RenderableUiCase {
  view(): string;
}

interface PortalRootProps {
  registry: PortalRegistry;
  createUiContext: (extra?: Record<string, unknown>) => UiContext;
  taskCreateUi?: RenderableUiCase;
  flash?:
    | {
        type: "success" | "error";
        message: string;
      }
    | null;
  moveError?:
    | {
        taskId: string;
        message: string;
      }
    | null;
}

export async function PortalRoot(props: PortalRootProps): Promise<string> {
  const DesignSystem = props.registry._packages.designSystem;
  const TaskCreateUi = props.registry._cases.tasks.task_create.ui;
  const TaskListUi = props.registry._cases.tasks.task_list.ui;
  const TaskMoveUi = props.registry._cases.tasks.task_move.ui;
  const taskCreateUi =
    props.taskCreateUi ??
    new TaskCreateUi(
      props.createUiContext({
        actionPath: "/actions/task-create",
        dialogId: "create-task-dialog",
      })
    );

  const taskListUi = new TaskListUi(
    props.createUiContext({
      renderCardActions: (task: PortalTask) => {
        return new TaskMoveUi(
          props.createUiContext({
            task,
            actionPath: "/actions/task-move",
            errorMessage:
              props.moveError?.taskId === task.id
                ? props.moveError.message
                : undefined,
          })
        ).view();
      },
    })
  );

  await taskListUi.load({});

  const taskCreateView = taskCreateUi.view();
  const taskListView = taskListUi.view();

  return DesignSystem.renderDocument({
    title: "APP Node Example",
    body: DesignSystem.AppShell({
      title: "Task Board",
      subtitle:
        "Node APP example with create, list, and move wired through canonical Cases.",
      children: [
        props.flash ? DesignSystem.Notice(props.flash) : "",
        DesignSystem.BoardHeader({
          title: "Board",
          subtitle:
            "Tasks load from the backend and each card can move across columns through the UI surfaces.",
        }),
        taskCreateView,
        taskListView,
      ].join(""),
    }),
  });
}
